Spigot-Server-Patches/0622-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch

```diff
@@ -0,0 +1,356 @@
++            Block block = world.getType(position).getBlock();
++            if (block instanceof BlockChest) {
++                  return BlockChest.getInventory((BlockChest)block, block.getBlockData(), world, position,false);
++              }
++            if (block.isTileEntity()) {
++                  TileEntity tile = world.getTileEntity(position);
++                  if (tile instanceof IInventory) return (IInventory) tile;
++              }
++            return null;
++        }
++
++    }
+diff --git a/src/main/java/io/papermc/paper/HopperPusher.java b/src/main/java/io/papermc/paper/HopperPusher.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..f225bd1ba158d58e3973f1e720f40510101c263c
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/HopperPusher.java
+@@ -0,0 +1,74 @@
++package io.papermc.paper;
++
++import java.util.Map;
++
++import net.minecraft.server.AxisAlignedBB;
++import net.minecraft.server.BlockPosition;
++import net.minecraft.server.Entity;
++import net.minecraft.server.TileEntity;
++import net.minecraft.server.TileEntityHopper;
++import net.minecraft.server.World;
++
++public interface HopperPusher {
++
++    default TileEntityHopper findHopper() {
++        BlockPosition pos = new BlockPosition(locX(), locY(), locZ());
++        int startX = pos.getX() - 1;
++        int endX = pos.getX() + 1;
++        int startY = Math.max(0, pos.getY() - 1);
++        int endY = Math.min(255, pos.getY() + 1);
++        int startZ = pos.getZ() - 1;
++        int endZ = pos.getZ() + 1;
++        for (int x = startX; x <= endX; x++) {
++            for (int y = startY; y <= endY; y++) {
++                for (int z = startZ; z <= endZ; z++) {
++                    BlockPosition adjacentPos = new BlockPosition(x, y, z);
++                    TileEntityHopper hopper = HopperHelper.getHopper(getWorld(), adjacentPos);
++                    if (hopper == null) {
++                        continue; // Avoid playing with the bounding boxes, if at all possible
++                    }
++                    AxisAlignedBB hopperBoundingBox = hopper.getHopperLookupBoundingBox();
++                    if (hopperBoundingBox.intersects(this
++                        .getBoundingBox())) {
++                        return hopper;
++                    }
++                }
++            }
++        }
++        return null;
++    }
++
++    boolean acceptItem(TileEntityHopper hopper);
++
++
++    default boolean tryPutInHopper() {
++        if (!getWorld().paperConfig.isHopperPushBased) {
++            return false;
++        }
++        TileEntityHopper hopper = findHopper();
++        return hopper != null && hopper.canAcceptItems() && acceptItem(hopper);
++    }
++
++
++    AxisAlignedBB getBoundingBox();
++
++
++    World getWorld();
++
++    // Default implementations for entities
++
++
++    default double locX() {
++        return ((Entity) this).locX();
++    }
++
```

Comment:
also not a fan of these double new lines everywhere

Reply:
Dropped double new lines, fear of bad auto formatting.